File: scala-3/euler/src/main/scala/com/rbleggi/euler/P24.scala
package com.rbleggi.euler

@main def runP24(): Unit = {
  val digits = (0 to 9).toList
  var k = 1000000 - 1
  var available = digits
  var result = List.empty[Int]

  for (i <- 9 to 0 by -1) {
    val f = (1 to i).product
    val idx = if (f == 0) 0 else k / f
    result = result :+ available(idx)
    available = available.patch(idx, Nil, 1)
    k = k % f
  }

  println(s"The millionth lexicographic permutation is: ${result.mkString}")
}

File: kotlin/euler/src/main/kotlin/com/rbleggi/euler/P24.kt
package com.rbleggi.euler

fun mainP24() {
    val digits = (0..9).toList()
    var k = 1000000 - 1
    var available = digits.toMutableList()
    val result = mutableListOf<Int>()

    for (i in 9 downTo 0) {
        val f = (1..i).fold(1) { acc, v -> acc * v }
        val idx = if (f == 0) 0 else k / f
        result.add(available[idx])
        available.removeAt(idx)
        k %= f
    }

    println("The millionth lexicographic permutation is: ${result.joinToString("")}")
}
